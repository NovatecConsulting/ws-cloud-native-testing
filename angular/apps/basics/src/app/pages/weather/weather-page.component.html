<h1>Weather App Basic examples</h1>
<basics-weather-introduction class="block pb-4"></basics-weather-introduction>
<weather-page-template
  [weatherTemplateData]="weatherTemplateData"
  (savedLocation)="saveLocation($event)"
  (searched)="search($event)"
></weather-page-template>
<basics-weather-explanation class="block pb-4"></basics-weather-explanation>
