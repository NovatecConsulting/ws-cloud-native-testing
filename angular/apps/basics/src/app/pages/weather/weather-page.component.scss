@import 'libs/theming/src/styles/page';
