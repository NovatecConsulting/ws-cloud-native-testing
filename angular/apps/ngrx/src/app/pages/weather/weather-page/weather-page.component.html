<p>weather-page works!</p>
