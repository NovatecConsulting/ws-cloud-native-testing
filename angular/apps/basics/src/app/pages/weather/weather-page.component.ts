import { Component } from '@angular/core';
import { WeatherService } from './service/weather.service';
import { WeatherLocation } from '@cntws/weather';

@Component({
  selector: 'basics-weather-page',
  templateUrl: './weather-page.component.html',
  styleUrls: ['./weather-page.component.scss'],
})
export class WeatherPageComponent {
  weatherTemplateData = {
    weather$: this.service.weather$,
    warning$: this.service.warning$,
    isLoading$: this.service.isLoading$,
    mainLocation$: this.service.mainLocation$,
  };

  constructor(private service: WeatherService) {
    this.service.getMainLocation();
    this.weatherTemplateData.mainLocation$.subscribe((location) => {
      if (location) {
        this.service.getWeatherForLocation(location);
      }
    });
  }

  search(location: string) {
    this.service.getWeatherForLocation(location);
  }

  saveLocation(location: WeatherLocation) {
    this.service.saveMainLocation(location.location);
  }
}
