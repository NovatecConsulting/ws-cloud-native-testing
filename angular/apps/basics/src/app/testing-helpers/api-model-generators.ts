import { LocationApiModel, WeatherApiModel } from '../pages/weather/service/weather.service';

const createWeatherApiModel: (name: string, temp: number) => WeatherApiModel[] = (name: string, temp: number) => {
  return [{ temp, name }];
};

const createLocationApiModel: (name: string) => LocationApiModel = (name: string) => {
  return { name };
};

export const ApiModelGenerators = {
  createWeatherApiModel,
  createLocationApiModel,
};
