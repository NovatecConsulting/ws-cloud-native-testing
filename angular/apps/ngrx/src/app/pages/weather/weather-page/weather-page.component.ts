import { Component, OnInit } from '@angular/core';
import {WeatherFacade} from "../+state/weather.facade";

@Component({
  selector: 'ngrx-weather-page',
  templateUrl: './weather-page.component.html',
  styleUrls: ['./weather-page.component.scss'],
})
export class WeatherPageComponent implements OnInit {
  constructor(private weather: WeatherFacade) {}

  ngOnInit(): void {
    this.weather.init();
  }
}
