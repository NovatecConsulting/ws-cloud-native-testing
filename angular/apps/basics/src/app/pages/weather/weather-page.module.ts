import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WeatherPageComponent } from './weather-page.component';
import { WeatherService } from './service/weather.service';
import { ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatInputModule } from '@angular/material/input';
import { MatCardModule } from '@angular/material/card';
import { MatDividerModule } from '@angular/material/divider';
import { WeatherIntroductionComponent } from './paragraphs/introduction/weather-introduction.component';
import { WeatherExplanationComponent } from './paragraphs/explanation/weather-explanation.component';
import { MatIconModule } from '@angular/material/icon';
import { MatProgressSpinnerModule } from '@angular/material/progress-spinner';
import { MatExpansionModule } from '@angular/material/expansion';
import { WeatherModule } from '@cntws/weather';

@NgModule({
  declarations: [WeatherPageComponent, WeatherIntroductionComponent, WeatherExplanationComponent],
  imports: [
    CommonModule,
    ReactiveFormsModule,
    MatButtonModule,
    MatInputModule,
    MatCardModule,
    MatDividerModule,
    MatIconModule,
    MatProgressSpinnerModule,
    MatExpansionModule,
    WeatherModule,
  ],
  providers: [WeatherService],
})
export class WeatherPageModule {}
