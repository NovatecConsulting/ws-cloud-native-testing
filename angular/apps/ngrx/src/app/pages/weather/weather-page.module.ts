import { NgModule } from '@angular/core';
import { WeatherPageComponent } from './weather-page/weather-page.component';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import * as fromWeather from './+state/weather.reducer';
import { WeatherEffects } from './+state/weather.effects';
import { WeatherFacade } from './+state/weather.facade';

@NgModule({
  declarations: [WeatherPageComponent],
  providers: [
    WeatherFacade,
  ],
  imports: [
    StoreModule.forFeature(fromWeather.WEATHER_FEATURE_KEY, fromWeather.reducer),
    EffectsModule.forFeature([WeatherEffects])
  ],
  exports: [
    WeatherPageComponent
  ]
})
export class WeatherPageModule {}
