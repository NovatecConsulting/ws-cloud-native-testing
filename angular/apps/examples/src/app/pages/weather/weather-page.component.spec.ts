import '../../testing-helpers/window.mock';
import { MockBuilder, MockRender, MockService } from 'ng-mocks';
import { WeatherPageComponent } from './weather-page.component';
import { WeatherModule } from './weather.module';
import { WeatherLocation, WeatherService } from './service/weather.service';
import { of } from 'rxjs';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
import { HarnessLoader } from '@angular/cdk/testing';
import { TestbedHarnessEnvironment } from '@angular/cdk/testing/testbed';
import { environment } from '../../../environments/environment';
import { ApiModelGenerators } from '../../testing-helpers/api-model-generators';
import { MatCardHarness } from '@angular/material/card/testing';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { MatInputHarness } from '@angular/material/input/testing';
import { MatButtonHarness } from '@angular/material/button/testing';

type serviceMockProps = {
  isLoading?: boolean;
  warning?: string | undefined;
  weather?: WeatherLocation[];
  mainLocation?: string | undefined;
};

function createServiceMock(opts: serviceMockProps) {
  const stateOverrides = {
    isLoading$: of(!!opts.isLoading),
    warning$: of(opts.warning),
    weather$: of(opts.weather ? opts.weather : []),
    mainLocation$: of(opts.mainLocation),
  };
  return MockService(WeatherService, stateOverrides);
}

describe('A user visiting the Component Testing Page', () => {
  /**
   * These tests are testing the complete module and give us confidence for bigger refactorings.
   * They are rather brittle and thus, we have to use them carefully.
   * Further, they do not give good insights in why a specific scenario is failing.
   * This loads to a need of additional tests for your components to get better feedback.
   */
  describe('and has his data loaded', () => {
    let fixture: ComponentFixture<WeatherPageComponent>;
    let loader: HarnessLoader;
    let controller: HttpTestingController;

    beforeEach(async () => {
      return TestBed.configureTestingModule({
        imports: [NoopAnimationsModule, WeatherModule, HttpClientTestingModule],
      }).compileComponents();
    });

    beforeEach(() => {
      fixture = TestBed.createComponent(WeatherPageComponent);
      loader = TestbedHarnessEnvironment.loader(fixture);
      controller = TestBed.inject(HttpTestingController);
    });

    it('should load his saved location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      await loader.getHarness(MatCardHarness.with({ subtitle: new RegExp(`.*(${location}).*`) }));
    });

    it('should load the weather for his saved location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      await loader.getHarness(MatCardHarness.with({ title: location }));
    });

    it('should be able to search for a location', async () => {
      const location = 'stuttgart';
      const temp = 25;

      const locationRequest = controller.expectOne(`${environment.weatherApi}/mainLocation`);
      locationRequest.flush(ApiModelGenerators.createLocationApiModel(location));
      const weatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${location}`);
      weatherRequest.flush(ApiModelGenerators.createWeatherApiModel(location, temp));

      const locationCard = await loader.getHarness(MatCardHarness.with({ subtitle: new RegExp(`.*(location).*`) }));
      const locationInput = await locationCard.getHarness(MatInputHarness);
      const searchedWeather = 'Frankfurt';
      await locationInput.setValue(searchedWeather);
      const searchButton = await locationCard.getHarness(MatButtonHarness);
      await searchButton.click();
      const searchedWeatherRequest = controller.expectOne(`${environment.weatherApi}/locations?q=${searchedWeather}`);
      searchedWeatherRequest.flush(ApiModelGenerators.createWeatherApiModel(searchedWeather, temp));

      await loader.getHarness(MatCardHarness.with({ title: searchedWeather }));
    });
  });

  describe('should have a consistent layout', () => {
    beforeEach(() => MockBuilder(WeatherPageComponent, WeatherModule));

    it('if the service is loading', () => {
      const service = createServiceMock({ isLoading: true });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherService, useValue: service }] });
      expect(fixture).toMatchSnapshot();
    });

    it('if the service has a warning', () => {
      const service = createServiceMock({ warning: 'This is a warning!' });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherService, useValue: service }] });
      expect(fixture).toMatchSnapshot();
    });

    it('if the service has no warning', () => {
      const service = createServiceMock({
        weather: [
          { location: 'Stuttgart', temp: 23 },
          { location: 'Frankfurt', temp: 15 },
        ],
      });
      const fixture = MockRender(WeatherPageComponent, {}, { providers: [{ provide: WeatherService, useValue: service }] });
      expect(fixture).toMatchSnapshot();
    });
  });
});
