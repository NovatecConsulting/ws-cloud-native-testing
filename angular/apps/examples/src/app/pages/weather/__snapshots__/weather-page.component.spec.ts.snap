// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A user visiting the Component Testing Page should have a consistent layout if the service has a warning 1`] = `
<mock-render
  __ngContext__={[Function LRootView]}
>
  <examples-weather-page>
    <h1>
      Weather App Basic examples
    </h1>
    <examples-weather-introduction
      class="block pb-4"
    />
    <div
      class="block relative min-h-100"
    >
      <examples-location
        class="block pb-4"
      />
      <examples-weather-results
        class="block pb-4"
      />
      <examples-error-message
        class="block mb-3"
      />
    </div>
    <examples-weather-explanation
      class="block pb-4"
    />
  </examples-weather-page>
</mock-render>
`;

exports[`A user visiting the Component Testing Page should have a consistent layout if the service has no warning 1`] = `
<mock-render
  __ngContext__={[Function LRootView]}
>
  <examples-weather-page>
    <h1>
      Weather App Basic examples
    </h1>
    <examples-weather-introduction
      class="block pb-4"
    />
    <div
      class="block relative min-h-100"
    >
      <examples-location
        class="block pb-4"
      />
      <examples-weather-results
        class="block pb-4"
      />
    </div>
    <examples-weather-explanation
      class="block pb-4"
    />
  </examples-weather-page>
</mock-render>
`;

exports[`A user visiting the Component Testing Page should have a consistent layout if the service is loading 1`] = `
<mock-render
  __ngContext__={[Function LRootView]}
>
  <examples-weather-page>
    <h1>
      Weather App Basic examples
    </h1>
    <examples-weather-introduction
      class="block pb-4"
    />
    <div
      class="block relative min-h-100"
    >
      <examples-loading />
    </div>
    <examples-weather-explanation
      class="block pb-4"
    />
  </examples-weather-page>
</mock-render>
`;
